src/stores/feature-flag.ts

```diff
@@ -0,0 +1,115 @@
+import { defineStore } from 'pinia'
+
+import featureData from '~~/feat/feature-flags.json'
+
+import { warn } from '~/utils/console'
+
+import type { FeatureFlag } from '~/types/feature-flag'
+import type { FeatureState, FlagStatus } from '~/constants/feature-flag'
+import {
+  ENABLED,
+  SWITCHABLE,
+  ON,
+  OFF,
+  DISABLED,
+} from '~/constants/feature-flag'
+import { LOCAL, DEPLOY_ENVS, DeployEnv } from '~/constants/deploy-env'
+
+export interface FeatureFlagState {
+  flags: Record<keyof typeof featureData['features'], FeatureFlag>
+}
+
+const FEATURE_FLAG = 'feature_flag'
+
+/**
+ * Get the status of the flag. If the flag status is environment dependent, this
+ * function will use the flag status for the current environment based on the
+ * `DEPLOYMENT_ENV` environment variable.
+ *
+ * @param flag - the flag for which to get the status
+ */
+export const getFlagStatus = (flag: FeatureFlag): FlagStatus => {
+  const deployEnv = (process.env.DEPLOYMENT_ENV ?? LOCAL) as DeployEnv
+  if (typeof flag.status === 'string') return flag.status
+  else {
+    const envIndex = DEPLOY_ENVS.indexOf(deployEnv)
+    for (let i = envIndex; i < DEPLOY_ENVS.length; i += 1) {
+      if (DEPLOY_ENVS[i] in flag.status) return flag.status[DEPLOY_ENVS[i]]
+    }
+  }
+  return DISABLED
+}
+
+/**
+ * Get the state of the feature based on the status of the feature flag and the
+ * preferences of the user.
+ *
+ * @param flag - the flag for which to get the state
+ */
+const getFeatureState = (flag: FeatureFlag): FeatureState => {
+  const status = getFlagStatus(flag)
+  if (status === SWITCHABLE)
+    return flag.preferredState ?? flag.defaultState ?? OFF
+  if (status === ENABLED) return ON
+  return OFF
+}
+
+export const useFeatureFlagStore = defineStore(FEATURE_FLAG, {
+  state: () =>
+    ({
+      flags: featureData.features,
+    } as FeatureFlagState),
+  getters: {
+    /**
+     * Get the state of the named feature, based on config and cookie.
+     */
+    featureState:
+      (state: FeatureFlagState) =>
+      (name: keyof typeof featureData['features']): FeatureState => {
+        if (name in state.flags) return getFeatureState(state.flags[name])
+        else {
+          warn(`Invalid feature flag accessed: ${name}`)
+          return ON
```

Comment:
Wouldn't it be safer to return OFF for an invalid feature?

Reply:
That's a good point, but returning `ON` by default allows us to drop the flags from the file and ensure a rolled-out feature does not get rolled back.